contracts/protocols/mainnet/euler/helpers.sol

```diff
@@ -0,0 +1,235 @@
+        uint256 supplyAPY;
+        // Pricing
+        uint256 twap;
+        uint256 currPrice;
+        // Account specific
+        uint256 underlyingBalance;
+        uint256 eulerAllowance;
+        uint256 eTokenBalance;
+        uint256 eTokenBalanceUnderlying;
+        uint256 dTokenBalance;
+        uint256 eTokenPriceUSD;
+        uint256 dTokenPriceUSD;
+        uint256 riskAdjustedColValue;
+        uint256 riskAdjustedBorValue;
+        uint256 riskAdjustedCol;
+        uint256 riskAdjustedLiability;
+        uint256 numBorrows;
+    }
+
+    /**
+     * @dev Return ethereum address
+     */
+    function getEthAddr() internal pure returns (address) {
+        return 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE; // ETH Address
+    }
+
+    /**
+     * @dev Return Weth address
+     */
+    function getWethAddr() internal pure returns (address) {
+        return 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2; // Mainnet WETH Address
+        // return 0xd0A1E359811322d97991E03f863a0C30C2cF029C; // Kovan WETH Address
+    }
+
+    /**
+     * @dev Get all sub-accounts of a user.
+     * @notice Get all sub-accounts of a user.
+     * @param user Address of user
+     */
+    function getAllSubAccounts(address user) public pure returns (SubAccount[] memory subAccounts) {
+        uint256 length = 256;
+        subAccounts = new SubAccount[](length);
+
+        for (uint256 i = 0; i < length; i++) {
+            address subAccount = getSubAccountAddress(user, i);
+            subAccounts[i] = SubAccount({ id: i, subAccountAddress: subAccount });
+        }
+    }
+
+    /**
+     * @dev Get all sub-accounts of a user.
+     * @notice Get all sub-accounts of a user.
+     * @param primary Address of user
+     * @param subAccountId sub-account Id(0 for primary and 1 - 255 for sub-account)
+     */
+    function getSubAccountAddress(address primary, uint256 subAccountId) public pure returns (address) {
+        require(subAccountId < 256, "sub-account-id-too-big");
+        return address(uint160(primary) ^ uint160(subAccountId));
+    }
+
+    /**
+     * @dev Get active sub-accounts.
+     * @notice Get active sub-accounts.
+     * @param subAccounts Array of SubAccount struct(id and address)
+     * @param tokens Array of the tokens
+     */
+    function getActiveSubAccounts(SubAccount[] memory subAccounts, address[] memory tokens)
+        public
+        view
+        returns (bool[] memory activeSubAcc, uint256 count)
+    {
+        uint256 accLength = subAccounts.length;
+        uint256 tokenLength = tokens.length;
+        activeSubAcc = new bool[](accLength);
+
+        for (uint256 i = 0; i < accLength; i++) {
+            for (uint256 j = 0; j < tokenLength; j++) {
+                address eToken = markets.underlyingToEToken(tokens[j]);
+
+                if ((IEToken(eToken).balanceOfUnderlying(subAccounts[i].subAccountAddress)) > 0) {
+                    activeSubAcc[i] = true;
+                    count++;
```

Comment:
Count can be equal to `activeSubAcc .length` right?

Reply:
No, the `activeSubAcc` is a list of bool values of all the subAccounts, so we need a count of all the `true` values.

Reply:
Ohh Alright